ariadne/src/Ariadne/Wallet/Backend/Tx.hs

```diff
@@ -74,22 +77,55 @@ sendTx ::
     -> InputSelectionPolicy
     -> NonEmpty TxOut
     -> IO TxId
-sendTx acidDb WalletFace {..} CardanoFace {..} walletSelRef printAction pp walletRef _accRefs isp outs = do
+sendTx acidDb WalletFace {..} CardanoFace {..} walletSelRef printAction pp walletRef accRefs isp outs = do
     let Nat runCardanoMode = cardanoRunCardanoMode
     walletDb <- query acidDb Snapshot
     let wallets = walletDb ^. dbHdWallets
-    walletRootId <- resolveWalletRef walletSelRef walletRef walletDb
-    runCardanoMode $ sendTxDo wallets walletRootId =<< cardanoGetDiffusion
+    (walletRootId, walletAccounts) <-
+        resolveWalletRefThenRead walletSelRef walletRef
+        walletDb readAccountsByRootId
+    accountIds <- resolveAccountRefs walletRootId walletAccounts
+    let filteredAccounts :: IxSet HdAccount
+        filteredAccounts = maybe identity filterAccounts accountIds walletAccounts
+
+        filterAccounts :: NonEmpty HdAccountId -> IxSet HdAccount -> IxSet HdAccount
+        filterAccounts ids accounts = accounts @+ toList ids
+
+    runCardanoMode $
+        sendTxDo wallets walletRootId filteredAccounts =<< cardanoGetDiffusion
   where
-    sendTxDo :: HdWallets -> HdRootId -> Diffusion CardanoMode -> CardanoMode TxId
-    sendTxDo wallets walletRootId diffusion = do
-        let
-            walletAccounts :: IxSet HdAccount
-            walletAccounts = fromRight
-                -- Resolved walletRootId always exists.
-                (error "SendTx: Unknown walletRootId")
-                (readAccountsByRootId walletRootId wallets)
-
+    -- Returns list of accounts which can be used.
+    -- 'Nothing' means all accounts can be used.
+    resolveAccountRefs ::
+           HdRootId
+        -> IxSet HdAccount
+        -> IO (Maybe (NonEmpty HdAccountId))
+    resolveAccountRefs rootId accountsSet =
+        case nonEmpty accRefs of
+            Nothing -> do
+                readIORef walletSelRef >>= \case
+                    Just (WSAccount selectedAccId)
+                        | selectedAccId ^. hdAccountIdParent == rootId ->
+                            return (Just $ one selectedAccId)
+                    _ -> return Nothing
+            Just accRefsNE -> Just <$> mapM refToId accRefsNE
+      where
+        -- Here we use knowledge of the order in UI, which is not very good.
```

Comment:
I'm not even sure how the order of `IxSet` corresponds to the UI order. Could you explain that part?

Reply:
We have implicit contract that UI indices should be treated as indices in `toList @(IxSet HdSomething)`. I find it fragile and want to refactor it. I think that backend shouldn't know about UI indices at all and all commands should take stable identifiers. So `LocalAccountReference` should have a constructor with `HdAccountIx`, not `Word`. Then we'll be able to do lookup using primary key, not using `toList` and `ix`.

Reply:
Hm, I thought it would be fixed by AD-168, but in fact AD-168 can fix it only for `select`, not for commands like `new-account` which use `tyWalletRef`.
I'll create a separate issue for this.

Reply:
https://issues.serokell.io/issue/AD-294